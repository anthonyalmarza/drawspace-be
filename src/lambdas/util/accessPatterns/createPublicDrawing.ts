import drawingEntity from 'lambdas/util/entities/drawing'
import { Drawing, UpdateReturnValues } from 'types'
import ksuid from 'ksuid'
import dayjs from 'dayjs'

export default async ({
    id = ksuid.randomSync().string,
    title,
    thumbnailUrl,
    drawStepsUrl,
    startTime,
    endTime,
    width,
    height,
    resolution,
    user,
}: {
    id?: string
    title: string
    thumbnailUrl: string
    drawStepsUrl: string
    startTime: string
    endTime: string
    width: number
    height: number
    resolution: number
    user: string
}): Promise<Drawing> => {
    const dayjsNow = dayjs()
    const { Attributes: drawing } = await drawingEntity.update(
        {
            id,
            title,
            thumbnailUrl,
            drawStepsUrl,
            startTime,
            endTime,
            width,
            height,
            resolution,
            user,
            published: dayjsNow.toISOString(),
            gsi1pk: dayjsNow.format('YYYY-MM'),
            gsi1sk: dayjsNow.toISOString(),
        },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return drawing
}
