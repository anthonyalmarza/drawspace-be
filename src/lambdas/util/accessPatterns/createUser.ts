import userEntity from 'lambdas/util/entities/user'
import { User, UpdateReturnValues } from 'types'

export default async ({
    userId,
    name,
}: {
    userId: string
    name: string
}): Promise<User> => {
    const { Attributes: user } = await userEntity.update(
        { id: userId, name },
        {
            returnValues: UpdateReturnValues.ALL_NEW,
            conditions: { attr: 'pk', exists: false },
        }
    )
    return user
}
