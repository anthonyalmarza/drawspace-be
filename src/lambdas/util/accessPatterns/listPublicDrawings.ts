import last from 'ramda/src/last'

import drawingEntity from 'lambdas/util/entities/drawing'
import fetchTruncatedTimestamp from 'lambdas/util/fetchTruncatedTimestamp'

import { Drawing, LastEvaluatedKey, DdbIndex } from 'types'

type PublicDrawings = {
    items: Drawing[]
    lastEvaluatedKey?: LastEvaluatedKey
}

export default (startKey?: LastEvaluatedKey): Promise<PublicDrawings> =>
    fetchTruncatedTimestamp<Drawing>({
        limit: 25,
        oldestDateInclusive: '2020-10',
        makeQueryPkFn: (fetchMonth) => `gsi1pkDrawing#${fetchMonth}`,
        index: DdbIndex.GSI1,
        whichPk: 'gsi1pk',
        beginsWith: 'gsi1skDrawing#',
        entity: drawingEntity,
        parseDateFromPkFn: (pk) => {
            const parsedDate = last(pk.split('#'))
            if (parsedDate) {
                return parsedDate
            }
            throw new Error(`Couldnt parse date from pk: ${pk}`)
        },
        startKey,
    })
