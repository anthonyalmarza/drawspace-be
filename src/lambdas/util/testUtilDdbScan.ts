import DynamoDB from 'aws-sdk/clients/dynamodb'

const { apiTableName } = require('lambdas/util/envVarsRestApi')
/* eslint-enable */
const DocumentClient = new DynamoDB.DocumentClient({
    endpoint: 'http://localhost:9001',
    accessKeyId: 'dynamo',
    secretAccessKey: 'devDummyKey',
    region: 'local',
})

export default () => DocumentClient.scan({ TableName: apiTableName }).promise()
