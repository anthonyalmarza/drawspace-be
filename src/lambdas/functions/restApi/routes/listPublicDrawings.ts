import { API, HandlerFunction } from 'lambda-api'
import validateRequest from 'lambdas/functions/restApi/middleware/validateRequest'
import listPublicDrawings from 'lambdas/util/accessPatterns/listPublicDrawings'
import { parseStartKey } from 'lambdas/util/pagination'

const validate = validateRequest(
    {
        startKey: { type: 'string' },
    },
    'query'
)

const handler: HandlerFunction = async (req) => {
    const { startKey: b64StartKey } = req.query
    const parsedStartKey = parseStartKey(b64StartKey)
    const res = await listPublicDrawings(parsedStartKey)
    return res
}

export default (api: API): void => {
    api.get('/drawings', validate, handler)
}
