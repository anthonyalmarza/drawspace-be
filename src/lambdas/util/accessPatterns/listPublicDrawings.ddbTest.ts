import { advanceTo } from 'jest-date-mock'
import testUtilDdbScan from 'lambdas/util/testUtilDdbScan'
import createPublicDrawing from 'lambdas/util/accessPatterns/createPublicDrawing'
import createUser from 'lambdas/util/accessPatterns/createUser'
import listPublicDrawings from 'lambdas/util/accessPatterns/listPublicDrawings'

const mockUserIds = ['1', '2']

test('listPublicDrawings', async () => {
    const drawingMockCommon = {
        title: 'mockTitle',
        thumbnailUrl: 'mockThumbnailUrl',
        drawStepsUrl: 'mockDrawStepsUrl',
        startTime: 'mockStartTime',
        endTime: 'mockEndTime',
        width: 1000,
        height: 1000,
        resolution: 2,
    }
    advanceTo(new Date(2020, 10, 10, 0, 0, 0))
    await Promise.all([
        createUser({ userId: mockUserIds[0], name: 'mock user name 1' }),
        createUser({ userId: mockUserIds[1], name: 'mock user name 2' }),
    ])
    await Array.from({ length: 10 }, (x, i) => i).reduce(async (p, id) => {
        await p
        advanceTo(new Date(2020, 9, 31, id, 0, 0))
        return createPublicDrawing({
            id: id.toString(),
            user: mockUserIds[id % 2 === 0 ? 0 : 1],
            ...drawingMockCommon,
        })
    }, Promise.resolve())
    const scan = await testUtilDdbScan()
    const res = await listPublicDrawings()
    expect(res).toMatchSnapshot()
    expect(scan).toMatchSnapshot()
})
