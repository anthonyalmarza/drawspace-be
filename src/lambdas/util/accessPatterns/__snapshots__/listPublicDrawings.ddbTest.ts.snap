// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listPublicDrawings 1`] = `
Object {
  "items": Array [
    Object {
      "created": "2020-10-31T09:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "9",
      "modified": "2020-10-31T09:00:00.000Z",
      "published": "2020-10-31T09:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "2",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T08:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "8",
      "modified": "2020-10-31T08:00:00.000Z",
      "published": "2020-10-31T08:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "1",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T07:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "7",
      "modified": "2020-10-31T07:00:00.000Z",
      "published": "2020-10-31T07:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "2",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T06:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "6",
      "modified": "2020-10-31T06:00:00.000Z",
      "published": "2020-10-31T06:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "1",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T05:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "5",
      "modified": "2020-10-31T05:00:00.000Z",
      "published": "2020-10-31T05:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "2",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T04:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "4",
      "modified": "2020-10-31T04:00:00.000Z",
      "published": "2020-10-31T04:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "1",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T03:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "3",
      "modified": "2020-10-31T03:00:00.000Z",
      "published": "2020-10-31T03:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "2",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T02:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "2",
      "modified": "2020-10-31T02:00:00.000Z",
      "published": "2020-10-31T02:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "1",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T01:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "1",
      "modified": "2020-10-31T01:00:00.000Z",
      "published": "2020-10-31T01:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "2",
      "width": 1000,
    },
    Object {
      "created": "2020-10-31T00:00:00.000Z",
      "drawStepsUrl": "mockDrawStepsUrl",
      "endTime": "mockEndTime",
      "entity": "drawing",
      "height": 1000,
      "id": "0",
      "modified": "2020-10-31T00:00:00.000Z",
      "published": "2020-10-31T00:00:00.000Z",
      "resolution": 2,
      "startTime": "mockStartTime",
      "thumbnailUrl": "mockThumbnailUrl",
      "title": "mockTitle",
      "user": "1",
      "width": 1000,
    },
  ],
  "lastEvaluatedKey": undefined,
}
`;

exports[`listPublicDrawings 2`] = `
Object {
  "Count": 12,
  "Items": Array [
    Object {
      "_ct": "2020-10-31T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T00:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T00:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T00:00:00.000Z",
      "pk": "drawing#0",
      "r": 2,
      "sk": "drawing#0",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "1",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T08:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T08:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T08:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T08:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T08:00:00.000Z",
      "pk": "drawing#8",
      "r": 2,
      "sk": "drawing#8",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "1",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T02:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T02:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T02:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T02:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T02:00:00.000Z",
      "pk": "drawing#2",
      "r": 2,
      "sk": "drawing#2",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "1",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T05:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T05:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T05:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#2",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T05:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T05:00:00.000Z",
      "pk": "drawing#5",
      "r": 2,
      "sk": "drawing#5",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "2",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T09:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T09:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T09:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#2",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T09:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T09:00:00.000Z",
      "pk": "drawing#9",
      "r": 2,
      "sk": "drawing#9",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "2",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-11-10T00:00:00.000Z",
      "_et": "user",
      "_md": "2020-11-10T00:00:00.000Z",
      "dc": 0,
      "n": "mock user name 2",
      "pdc": 0,
      "pk": "user#2",
      "sk": "user#2",
    },
    Object {
      "_ct": "2020-10-31T04:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T04:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T04:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T04:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T04:00:00.000Z",
      "pk": "drawing#4",
      "r": 2,
      "sk": "drawing#4",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "1",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T07:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T07:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T07:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#2",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T07:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T07:00:00.000Z",
      "pk": "drawing#7",
      "r": 2,
      "sk": "drawing#7",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "2",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T06:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T06:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T06:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T06:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T06:00:00.000Z",
      "pk": "drawing#6",
      "r": 2,
      "sk": "drawing#6",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "1",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-10-31T03:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T03:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T03:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#2",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T03:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T03:00:00.000Z",
      "pk": "drawing#3",
      "r": 2,
      "sk": "drawing#3",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "2",
      "wi": 1000,
    },
    Object {
      "_ct": "2020-11-10T00:00:00.000Z",
      "_et": "user",
      "_md": "2020-11-10T00:00:00.000Z",
      "dc": 0,
      "n": "mock user name 1",
      "pdc": 0,
      "pk": "user#1",
      "sk": "user#1",
    },
    Object {
      "_ct": "2020-10-31T01:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-10-31T01:00:00.000Z",
      "d": "mockTitle",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-10",
      "gsi1sk": "gsi1skDrawing#2020-10-31T01:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#2",
      "gsi2sk": "gsi2skDrawing#public#2020-10-31T01:00:00.000Z",
      "he": 1000,
      "p": "2020-10-31T01:00:00.000Z",
      "pk": "drawing#1",
      "r": 2,
      "sk": "drawing#1",
      "st": "mockStartTime",
      "tu": "mockThumbnailUrl",
      "u": "2",
      "wi": 1000,
    },
  ],
  "ScannedCount": 12,
}
`;
